Modules/posixmodule.c

```diff
@@ -1834,17 +1837,110 @@ attributes_from_dir(LPCWSTR pszFile, BY_HANDLE_FILE_INFORMATION *info, ULONG *re
     return TRUE;
 }
 
+static void
+win32_xstat_fixup_exec_mode(const wchar_t *path,
+                            struct _Py_stat_struct *result)
+{
+    if (!(result->st_file_attributes & FILE_ATTRIBUTE_DIRECTORY)) {
+        /* Fix the file execute permissions. This hack sets S_IEXEC if
+           the filename has an extension that is commonly used by files
+           that CreateProcessW can execute. A real implementation calls
+           GetSecurityInfo, OpenThreadToken/OpenProcessToken, and
+           AccessCheck to check for generic read, write, and execute
+           access. */
+        const wchar_t *fileExtension = wcsrchr(path, '.');
+        if (fileExtension) {
+            if (_wcsicmp(fileExtension, L".exe") == 0 ||
+                _wcsicmp(fileExtension, L".bat") == 0 ||
+                _wcsicmp(fileExtension, L".cmd") == 0 ||
+                _wcsicmp(fileExtension, L".com") == 0) {
+                result->st_mode |= 0111;
+            }
+        }
+    }
+}
+
+static int
+win32_xstat_get_st_dev(const wchar_t *path,
+                       struct _Py_stat_struct *result)
+{
+    const wchar_t *rootEnd;
+    WCHAR rootBuffer[MAX_PATH];
+    WCHAR *root = rootBuffer;
+    DWORD oldErrorMode;
+    DWORD vsn = 0;
+    int retval = 0;
+
+    if (result->st_dev || !path || !path[0]) {
+        return 0;
+    }
+
+    if (PathCchSkipRoot(path, &rootEnd) ||
+        wcsncpy_s(rootBuffer, MAX_PATH, path, (rootEnd - path))) {
+        /* No root for the path, so let it use the current volume */
+        root = NULL;
+    }
```

Comment:
This is problematic. Instead, open a handle via `CreateFileW()` and call `GetVolumeInformationByHandleW()`. It's no more expensive than calling `GetVolumeInformationW()`, which also has to open a path, and it's reliable. That said, we're only here because it isn't a 'fast' stat, so we may as well take the slow path from the start.

`PathCchSkipRoot()` only supports backslash as the path separator. That's an irrelevant problem, however, considering the fundamental problem that the path may contain mount points, symlinks, and other reparse points. In this case the drive/device of the opened path is probably irrelevant to the file.

`GetVolumeInformationW()` will open any path that has a trailing slash. However, after opening the path it queries the `FileNameInformation` to verify that the final path ends in a backslash, i.e. that a filesystem root directory was opened. The problem is finding the relevant root directory in the path without resolving the final path. `GetVolumePathNameW()` is supposed to handle this, but I find that it's buggy with substitute drives, for which it returns ridiculous results. It also returns the wrong result for a symlink to a UNC path. And it's more work than simply opening the path. 

Actually, the requirement for a root directory is irrelevant to the underlying system call that queries the FileFsVolumeInformation (volume name and serial number) and FileFsAttributeInformation (filesystem name, flags, and max component length). Any open file on the volume is sufficient. Hence, just open the path and call `GetVolumeInformationByHandleW()`.

Reply:
Yes, I know all of this. I'm still having a private discussion with the Windows team about optimising `GetVolumeInformation` or adding VSN into the stat structure, either of which would let us bypass this extra step.

Right now, if someone passes `fast=False`, we could always just go back to the original slow path. But in the future, this slow path could get drastically faster without us having to change anything, which means automatic "backports". Right now it's about the same.

Reply:
> Yes, I know all of this. 

Okay, I just felt it was worth discussing since the result of `win32_xstat_get_st_dev()` as written would often be incorrect.

> optimising GetVolumeInformation 

They really should fix `GetVolumeInformationW()` to work consistently with `GetVolumeInformationByHandleW()`. All of the loops it jumps through to require a root directory are pointless and frustrating.

Reply:
> That's an irrelevant problem, however, considering the fundamental problem that the path may contain mount points, symlinks, and other reparse points.

I think we can probably define this as an edge case: "`lstat()` on Windows when passed a path including junctions or non-symlink reparse points that direct to another drive are ignored when getting `st_dev`. The device matching the start of the path will be returned, even if the reparse points mean the file is on another drive."

It's also possible that if the VSN gets added into the stat data before Windows releases, that we won't need this path at all.